serde/src/main/java/com/amazon/ionhiveserde/caseinsensitivedecorator/IonStructCaseInsensitiveDecorator.java

```diff
@@ -0,0 +1,132 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.ionhiveserde.caseinsensitivedecorator;
+
+import com.amazon.ion.ContainedValueException;
+import com.amazon.ion.IonStruct;
+import com.amazon.ion.IonValue;
+import com.amazon.ion.SymbolToken;
+import com.amazon.ion.UnknownSymbolException;
+import com.amazon.ion.ValueFactory;
+
+import java.util.Map;
+
+
+public class IonStructCaseInsensitiveDecorator extends IonContainerCaseInsensitiveDecorator implements IonStruct {
+    private final IonStruct ionStruct;
+
+    public IonStructCaseInsensitiveDecorator(final IonStruct s) {
+        super(s);
+        this.ionStruct = s;
+    }
+
+    @Override
+    public boolean containsKey(final Object fieldName) {
+        if (ionStruct.containsKey(fieldName)) {
+            return true;
+        } else {
+            for (IonValue v : ionStruct) {
+                if (fieldName instanceof String) {
+                    String s = (String) fieldName;
+                    if (s.equalsIgnoreCase(v.getFieldName())) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+    }
+
+    @Override
+    public boolean containsValue(final Object value) {
+        return ionStruct.containsValue(value);
+    }
+
+    @Override
+    public IonValue get(final String fieldName) {
+        if (ionStruct.containsKey(fieldName)) {
+            IonValue v = ionStruct.get(fieldName);
+            return IonCaseInsensitiveDecorator.wrapValue(v);
+        } else {
+            for (IonValue v : ionStruct) {
+                if (fieldName.equalsIgnoreCase(v.getFieldName())) {
+                    return IonCaseInsensitiveDecorator.wrapValue(v);
+                }
+            }
+            return ionStruct.get(fieldName);
+        }
+    }
+
+    @Override
+    public void put(final String fieldName, final IonValue child) throws ContainedValueException {
+        ionStruct.put(fieldName, child);
+    }
+
+    @Override
+    public ValueFactory put(final String fieldName) {
+        return ionStruct.put(fieldName);
+    }
+
+    @Override
+    public void putAll(final Map<? extends String, ? extends IonValue> m) {
+        ionStruct.putAll(m);
+    }
+
+    @Override
+    public void add(final String fieldName, final IonValue child) throws ContainedValueException {
+        ionStruct.add(fieldName, child);
+    }
+
+    @Override
+    public void add(final SymbolToken fieldName, final IonValue child) throws ContainedValueException {
+        ionStruct.add(fieldName, child);
+    }
+
+    @Override
+    public ValueFactory add(final String fieldName) {
+        return ionStruct.add(fieldName);
+    }
+
+    @Override
+    public IonValue remove(final String fieldName) {
+        IonValue v = ionStruct.remove(fieldName);
+        return IonCaseInsensitiveDecorator.wrapValue(v);
+    }
```

Comment:
Your milage may vary but I'd prefer a static import of `wrapValue` and the pattern below for cases like this:
```suggestion
    public IonValue remove(final String fieldName) {
        return wrapValue(ionStruct.remove(fieldName));
    }
```